qutebrowser/browser/browsertab.py

```diff
@@ -525,19 +525,73 @@ def at_bottom(self):
         raise NotImplementedError
 
 
+class TabHistoryItem:
+
+    """A single item in the tab history.
+
+    Attributes:
+        url: The QUrl of this item.
+        original_url: The QUrl of this item which was originally requested.
+        title: The title as string of this item.
+        active: Whether this item is the item currently navigated to.
+        user_data: The user data for this item.
+    """
+
+    def __init__(self, url, title, *, original_url=None, active=False,
+                 user_data=None):
+        self.url = url
+        if original_url is None:
+            self.original_url = url
+        else:
+            self.original_url = original_url
+        self.title = title
+        self.active = active
+        self.user_data = user_data
+
+    def __repr__(self):
+        return utils.get_repr(self, constructor=True, url=self.url,
+                              original_url=self.original_url, title=self.title,
+                              active=self.active, user_data=self.user_data)
+
+    @classmethod
+    def from_qt(cls, qt_item, active=False):
+        """Contstruct a TabHistoryItem from a Qt history item.
+
+        Args:
+            qt_item: a QWebEngineHistoryItem or QWebKitHistoryItem
+        """
+        qtutils.ensure_valid(qt_item)
+
+        try:
+            user_data = qt_item.userData()
+        except AttributeError:
+            user_data = None
+
+        return TabHistoryItem(
+            qt_item.url(),
+            qt_item.title(),
+            original_url=qt_item.originalUrl(),
+            active=active,
+            user_data=user_data)
+
+
 class AbstractHistory:
 
     """The history attribute of a AbstractTab."""
 
     def __init__(self, tab):
         self._tab = tab
         self._history = None
+        self.to_load = []
 
     def __len__(self):
         return len(self._history)
```

Comment:
Initially, `_history` was used but now i fully rely on `to_load` so that i only return instances of `AbstractTabHistoryItem`. I thought this was a better interface. I had to also hook into `on_load_finished` so that i add history items to `to_load` as the user navigates pages.